### ISSUE TITLE
MPE Controllers mapping and scaling

### ISSUE THREAD

**Issue by alfonso73** [surge-synthesizer/surge#2676]
**Is your feature request related to a problem? Please describe.**
When i design MPE based sounds i always need a way to map the response of the actual MPE data generated by the actual MPE device.

**Describe the solution you'd like**
I'd like to have on all 5 MPE dimensions (on velocity, off velocity, pitch bend, aftertouch, CC74) min, max range parameters and a mapping curve with the feature to insert points and handles for the segments curvature/linearity.
All this stuff can fit in the modulation area when the relevant controller source is selected (just as LFO are now)

**Comment by baconpaul** [comment]
I think your request is the same as this one: https://github.com/surge-synthesizer/surge/issues/1384 basically

I like the idea of using the modulation area to display mappers for fixed sources though; I do worry what we would do if we wanted to apply mappers to LFOs or MSEGs in that case.

Definitely on our radar though.

**Comment by alfonso73** [comment]
Yeah is similar. But i think that for MPE controllers is really needed to map the hardware generated data. #1384 is a more subtle thing and honestly i use it quite often in Zebra2/HZ. But it's another matter. Like adding some modulation processors (mappers, lag, inverters, etc.) wich can be used when you use the same mod source for different parameters but you need different response curves for the different modulated parameter. My request is more an MPE specific thing so i'd prefer a set of mappers and range operators in the relevant MPE sources tabs.

**Comment by mkruselj** [comment]
The idea **is** to have modmappers per target, not per source. Like in UVI Falcon, for example.

**Comment by baconpaul** [comment]
@mkruselj maybe we should consider a 'hardware calibrator' though (kinda like pianotec has per velocity calibration). A simple thing which as @alfonso73 says, could be displayed in the LFO tab for the hardware-inbound controllers to just apply a list of control points and a linear map. So basically an 'input scaler' more than a 'modulation mapper' which appies to the midi cc based modulators? I see how that's a bit different.

@alfonso73 - presume your request is motivated by, say, this section of the equator UI?

<img width="798" alt="Screen Shot 2020-09-23 at 5 26 15 PM" src="https://user-images.githubusercontent.com/13387561/94071455-e9212380-fdc1-11ea-92d6-cd834a0cc082.png">

**Comment by baconpaul** [comment]
Oh and we would of course do it for all midi inbound controllers, not just the MPE ones, but that's a minor detail.

**Comment by alfonso73** [comment]
@baconpaul yeah exactly a kind of hardware calibrator. In the line of Equator (i never used it though) or Cypher2 (i use it quite often. I think this is a "specialized" thing that can live along what @mkruselj says (mod mappers per target would be great indeed!)

**Comment by mkruselj** [comment]
Personally I think that sort of thing should be on the hardware device itself, I would rather go full-in with modmappers per target and we're golden for any sort of scenario. Bonus points for modmapper presets!

**Comment by alfonso73** [comment]
@mkruselj i humbly disagree beacause the main reason is that when you use an hardware controller with different sounds on the same virtual instrument or different virtual instruments would be an hassle to set the hardware internally for every sound. If you think about a live performance situation would be a nightmare. Anyway all the major and more mature MPE virtual instruments (Cypher2 is not by chance part of the ROLI offer) have a kind of mapping mechanisms of MPE data coming from hardware. I think it is not an option, is a real mandatory thing. When i desing sounds on Surge using MPE i constantly face this show-stopper

**Comment by mkruselj** [comment]
I have a Rise 49 myself, and I constantly find the need to adjust the behavior of each of the dimensions per patch. There is no one-size-fits all patches from my perspective...

**Comment by baconpaul** [comment]
Hold on - you two are disagreeing whether these mappers are per-patch or per-session?

I figured the would obviously be per-patch.

@alfonso73 you want it per session but just for surge?

**Comment by mkruselj** [comment]
No the way I understand it, Alfonso wants to have the mappers at the MIDI controller input level, whereas to me this is overly limiting (what if you want to use the same CC for another modulation with a different scaling?), and I'd rather have the mapper per target.

**Comment by baconpaul** [comment]
Right

So what I thought he was looking for (and he really should disambiguate for us) is

On (say) breath or timbre
when you click on the modulator
You can set a min/max/curvepoint type thing
and then in that patch, that CC gets transformed accordingly at midi input time

So it's basically a "how does 0-127 go to 0-1, on a per-patch basis, for each hardwired or custom midi bound modulator"

Which is exactly what the equator screen I showed does.

@alfonso73 ?

**Comment by alfonso73** [comment]
Hi all guys, i try to be as precise as i can with a practical case example. I'm in a live performance using a bank of 32 Surge patches that i designed to be used with my MPE controller of choice. What I need is that when i quickly change Surge patch to be used in a particular moment of the performance i need that all the MPE data are correctly mapped. I don't have time to change relevant hardware controller curves. So yes i'd need min/max and possibly some bezier curve to map the hardware generated data. This doesn't exculude having mappers per destination. This would even more flexibility which is obviously wellcome. As for MPE stuff i think we really need **at least** min/max and mapping lines/curves for relevant MPE data. If mappers will come for every destination...well this is even better. But at least adding in the MPE relevant tabs (as the LFO ones) some parameters for min/max and a curve/line feature is really needed.

**Comment by baconpaul** [comment]
Right I get that idea @alfonso73 - the question is: do you need this per patch? Or do you need this for all the surges you have open?

If you need it for all the surges, it should really be outside surge. 

If you need it for particular patches, then it's as you say a feature of the fixed input modulator.

Presume you set this up before the live performance and you are in MainStage or some such. The thing I don't get is the granularity you want this on.

**Comment by mkruselj** [comment]
For your use case indeed you should use external MIDI mappers to do that, it sounds like you want the same curves to apply globally.

**Comment by alfonso73** [comment]
Guys, totally per patch! Every patch need a different kind of mapping of the actual data coming from hardware. @mkruselj  nope man, if i have to change on the fly dozens of patches during a live set i need to get the right response for every different patch. **per patch**... probably my english is no so clear. Scuse me

**Comment by baconpaul** [comment]
Yeah OK gotcha. And it's not your english, it's our brains :)

So basically "exactly what equator has" is indeed what you want!

The CC mappers currently have a single model; and we have it our list to change that. (Right now the model is 'moderately smoothed exponential decay to linear application of target). We've been debating changing that over in #1412, and we definitely do want to have good defaults. But you make me think when I engineer #1412 I can do it at a patch level considering this feature in the future.

@mkruselj this is different from per-point mappers also, kinda, in that it really is an midi interpretation curve as posed. They are of course related but....

I will ponder some. I like this idea. I'm going to tentatively tag this issue 1.8.0 (which means we would target it for the january release, even though #1412 is already in that; this would be an addition) but may end up backing down on that depending on our workload. And part of the reason for that is I kinda want to use it on the sustain pedal midi controller for some patches I'm working on....

But anyway so we are clear, the full spec is

1. There's a per-patch re-mapping of any of the 'hard-mapped' controllers (which are the CCs, velocity, breath, expression, timbre, ATs, etc...)
2. The default of that remapping is "linear; with a smoothing function defaulted by #1412 but overridable per instance"
3. You can edit the transformation curve but it really is just a simple mseg type curve. You want to at least be able to do x -> -x; x -> ( 0 < 64; 2x > 64) and other type things. 
4. That transformation curve gets edited in the LFO section when you select a hard mapped modulator
5. That transformation saves with the patch

**Comment by alfonso73** [comment]
@baconpaul thanks for addition to the 1.8.0 milestone! Yes, it seems everything ok to me. I'd make possible to draw curve / lines and adding points exactly as an MSEG. And it should be possible to add points at coordinates like x = 0, y > 0 and x = 127 (or whatever the max value is even if it's normalized at 0..1 range), y < 127. This way minimum and maximum values coming from hardware can be mapped to a smaller range or to an inverted range etc.

**Comment by mkruselj** [comment]
@alfonso73 We have decided to move this issue to after 1.8.0 release. We still want to include it, but as a part of another bigger feature that will make it easier to implement all in one go. Plus, we currently have almost 50 open issues in the 1.8 milestone, and January is getting closer and closer 🙂 

I hope you can support us in this decision with patience. Thanks!

**Comment by baconpaul** [comment]
Yeah sorry @alfonso73 there's just too much on the milestone now and we really want to get it out (18 is gonna have loads of really useful stuff!) This one will make the top of the 19 list along with a patch browser is my bet.

Apologies.

**Comment by alfonso73** [comment]
No problem guys, i undersand. I'll just avoid MPE patching on Surge until that feature will land :-P Thanks for your great work!
